### ISSUE TITLE
Issue in ConfusionMatrix class in metrics.py

### ISSUE THREAD

**Issue by krishnakanagal** [ultralytics/yolov5#9813]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and found no similar bug report.

### YOLOv5 Component

Validation

### Bug

The confusion matrix calculation is wrong when there is no overlap between predictions and ground truth labels i.e. 0 IOU. In this scenario, the number of ground truth labels should be added as False Negatives and the predictions as False positives. But, the code does only the former. The False positives are not added to the confusion matrix hence leading to incorrect precision calculation. To illustrate this issue please refer to the example I have provided. I am ready to work on this pull request. 

### Environment

- YOLOv5 🚀 v6.1-266-g34df5032 Python-3.8.10 torch-1.10.0+cu102 CUDA:0 (Quadro RTX 8000, 48600MiB)
- Ubuntu
- python 3.8

### Minimal Reproducible Example

```

import torch
from metrics import ConfusionMatrix

# Initialize the confusion matrix with 2 classes
cm = ConfusionMatrix(2,0.25, 0.5)

actual = torch.tensor([[0.00000, 0.87695, 0.01367, 1.00000, 0.15430]], dtype=torch.float64)
predicted = torch.tensor([[0.37695, 0.37158, 0.62500, 0.47363, 0.51709, 0.00000]], dtype=torch.float64)

cm.process_batch(predicted, actual)
print(cm.matrix)
```
The output is this
[[0. 0. 0.]
 [0. 0. 0.]
 [1. 0. 0.]]
But, it should be this in this scenario
[[0. 0. 1.]
 [0. 0. 0.]
 [1. 0. 0.]]

### Additional

According to me, the fix is simple and I have given the fix for the code. Please let me know if you want me to submit PR.
```
class ConfusionMatrix:
    # Updated version of https://github.com/kaanakan/object_detection_confusion_matrix
    def __init__(self, nc, conf=0.25, iou_thres=0.45):
        self.matrix = np.zeros((nc + 1, nc + 1))
        self.nc = nc  # number of classes
        self.conf = conf
        self.iou_thres = iou_thres

    def process_batch(self, detections, labels):
        """
        Return intersection-over-union (Jaccard index) of boxes.
        Both sets of boxes are expected to be in (x1, y1, x2, y2) format.
        Arguments:
            detections (Array[N, 6]), x1, y1, x2, y2, conf, class
            labels (Array[M, 5]), class, x1, y1, x2, y2
        Returns:
            None, updates confusion matrix accordingly
        """
        detections = detections[detections[:, 4] > self.conf]
        gt_classes = labels[:, 0].int()
        detection_classes = detections[:, 5].int()
        iou = box_iou(labels[:, 1:], detections[:, :4])

        x = torch.where(iou > self.iou_thres)
        if x[0].shape[0]:
            matches = torch.cat((torch.stack(x, 1), iou[x[0], x[1]][:, None]), 1).cpu().numpy()
            if x[0].shape[0] > 1:
                matches = matches[matches[:, 2].argsort()[::-1]]
                matches = matches[np.unique(matches[:, 1], return_index=True)[1]]
                matches = matches[matches[:, 2].argsort()[::-1]]
                matches = matches[np.unique(matches[:, 0], return_index=True)[1]]
        else:
            matches = np.zeros((0, 3))

        n = matches.shape[0] > 0
        m0, m1, _ = matches.transpose().astype(int)
        for i, gc in enumerate(gt_classes):
            j = m0 == i
            if n and sum(j) == 1:
                self.matrix[detection_classes[m1[j]], gc] += 1  # correct
            else:
                self.matrix[self.nc, gc] += 1  # background FP

        if n:
            for i, dc in enumerate(detection_classes):
                if not any(m1 == i):
                    self.matrix[dc, self.nc] += 1  # background FN
        else: # (When there is no overlap in predicted and actual match is zero so the predicted should be added as false positive)
            for i, dc in enumerate(detection_classes):
                self.matrix[dc, self.nc] += 1  # background FP


    def matrix(self):
        return self.matrix

    def tp_fp(self):
        tp = self.matrix.diagonal()  # true positives
        fp = self.matrix.sum(1) - tp  # false positives
        # fn = self.matrix.sum(0) - tp  # false negatives (missed detections)
        return tp[:-1], fp[:-1]  # remove background class

    def plot(self, normalize=True, save_dir='', names=()):
        try:
            import seaborn as sn

            array = self.matrix / ((self.matrix.sum(0).reshape(1, -1) + 1E-9) if normalize else 1)  # normalize columns
            array[array < 0.005] = np.nan  # don't annotate (would appear as 0.00)

            fig = plt.figure(figsize=(12, 9), tight_layout=True)
            nc, nn = self.nc, len(names)  # number of classes, names
            sn.set(font_scale=1.0 if nc < 50 else 0.8)  # for label size
            labels = (0 < nn < 99) and (nn == nc)  # apply names to ticklabels
            with warnings.catch_warnings():
                warnings.simplefilter('ignore')  # suppress empty matrix RuntimeWarning: All-NaN slice encountered
                sn.heatmap(array,
                           annot=nc < 30,
                           annot_kws={
                               "size": 8},
                           cmap='Blues',
                           fmt='.2f',
                           square=True,
                           vmin=0.0,
                           xticklabels=names + ['background FP'] if labels else "auto",
                           yticklabels=names + ['background FN'] if labels else "auto").set_facecolor((1, 1, 1))
            fig.axes[0].set_xlabel('True')
            fig.axes[0].set_ylabel('Predicted')
            fig.savefig(Path(save_dir) / 'confusion_matrix.png', dpi=250)
            plt.close()
        except Exception as e:
            print(f'WARNING: ConfusionMatrix plot failure: {e}')

    def print(self):
        for i in range(self.nc + 1):
            print(' '.join(map(str, self.matrix[i])))
```

### Are you willing to submit a PR?

- [X] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@krishnakanagal can you please provide before and after results on VOC for example? i.e.
```
python val.py --data VOC.yaml
```

Dataset will auto-download on demand.

**Comment by krishnakanagal** [comment]
Also, the precision calculated from the confusion matrix for each class is not the same as the precision printed for each class on the console.

**Comment by hung-phan-van** [comment]
Same with you

**Comment by glenn-jocher** [comment]
> Also, the precision calculated from the confusion matrix for each class is not the same as the precision printed for each class on the console.

Yes this is true. This is because console printout is at max F1 confidence, while confusion matrix is produce at a fixed 0.25 confidence. 

I think to avoid complications we should simply align both at conf = 0.25
